src/Infrastructure/EventListener/OnFlushEntityObserverListener.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\PriceHistoryPlugin\Infrastructure\EventListener;
+
+use Doctrine\ORM\Event\OnFlushEventArgs;
+use Doctrine\ORM\UnitOfWork;
+use Sylius\PriceHistoryPlugin\Infrastructure\EntityObserver\EntityObserverInterface;
+use Webmozart\Assert\Assert;
+
+final class OnFlushEntityObserverListener
+{
+    public function __construct(private iterable $entityObservers)
+    {
+        Assert::allImplementsInterface($this->entityObservers, EntityObserverInterface::class);
+    }
+
+    public function onFlush(OnFlushEventArgs $eventArgs): void
+    {
+        $entityManager = $eventArgs->getObjectManager();
+        $unitOfWork = $entityManager->getUnitOfWork();
+
+        $scheduledEntities = array_merge(
+            $unitOfWork->getScheduledEntityInsertions(),
+            $unitOfWork->getScheduledEntityUpdates(),
+        );
+
+        foreach ($scheduledEntities as $entity) {
+            /** @var EntityObserverInterface $entityObserver */
+            foreach ($this->entityObservers as $entityObserver) {
+                $supportedEntity = $entityObserver->getSupportedEntity();
```

Comment:
Maybe you can refactor the observer to have `supports` method instead of `getSupportedEntity`, but it is only a suggestion according to the usage of this method

Reply:
Good idea, I've implemented that. Thanks.